### ISSUE TITLE
Why "additionalTextEdits" when one can send the whole document?

### ISSUE THREAD

**Issue by mickaelistria** [Microsoft/language-server-protocol#233]
I'm looking at the details of the CompletionItem, and I have to admit that I find the "additionalTextEdits" somehow complex to implement property. I'm wondering whether this is necessary in the protocol. It would seem much simpler to send the whole document or document subset in that case. Couldn't we simplify the protocol without reducing its abilities by getting rid of additionalTextEdits, or are there some known clients that heavily depend on it?

**Comment by nwolverson** [comment]
I'd imagine that sending the whole document instead of multiple edits could result in an incorrect cursor position - e.g. a completion which also adds an import to the top of the file, the cursor should move down to stay with the current line.

**Comment by kdvolder** [comment]
Yes, but you would be able to control precisely where the cursor ends up via `$0` snippet parameter.

This, in a way, is better than the current situation because, thanks to the restriction on the main edit's range, doing 'whole document' edit is not really possible at the moment (unless your document is only one line of text :-) So at the moment, its very hard to control precisely where the cursor ends up. You are severely constrained by the main edit's range limitation. So, for example its not possible to make a completion that inserts text in another part of your file and also moves your cursor there.

A good reason however to not have 'whole document edits' is to avoid transmitting lots of data. This is not as critical if you resolve completions lazily maybe, but is certainly problematic for eagerly resolved completions to send the entire document in every completion item.

**Comment by dbaeumer** [comment]
Actually the reason is to have a stable cursor position. Right now we don't support to change the cursor position during code complete (with the expection of snippets, but the cursor can only end inside the snippet). 

In addition minimal edits to a document have a couple of advantages:

- they are more performant
- they keep markers like breakpoints more stable. If you replace the whole document usually editors erase all additional markers since they don't know what to do with them unless the editor diffs the two contents and computes a minimal change again :-)

Rather than sending the whole document I would prefer having a feature request to control the cursor position.

Objection to close this ?

**Comment by mickaelistria** [comment]
> they keep markers like breakpoints more stable. If you replace the whole document usually editors erase all additional markers since they don't know what to do with them unless the editor diffs the two contents and computes a minimal change again :-)

I believe sync of document with markers (other that LSP diagnostics) isn't really in the scope of LSP. It's up to editors and the marker provider to decide of their strategy on this topic.

> Rather than sending the whole document I would prefer having a feature request to control the cursor position.
> Objection to close this ?

I think if we remove the restriction about edit line (#92 and #235), we can close the issue as it allows language servers to pick the strategy that's best for them (multi-line/whole document edit vs main edit on single line + additional edits).

**Comment by kdvolder** [comment]
> I believe sync of document with markers (other that LSP diagnostics) isn't really in the scope of LSP. It's up to editors and the marker provider to decide of their strategy on this topic.

Yes, but, whole document changes will make this harder. So I don't really get how that diminishes Dirk's explanation of why its good to have more fine-grained edits.

> I think if we remove the restriction about edit line (#92 and #235), we can close the issue as it allows language servers to pick the strategy that's best for them (multi-line/whole document edit vs main edit on single line + additional edits).

Hmmm yeah, I agree with that too. In the sense that this restriction looks somewhat arbitrary to me, and I don't really see a good reason for having it at all (but maybe Dirk can explain if there is some good reason after all :-)

On the other hand, I think, simply putting some thought into a way for the server to control cursor position really more directly addresses my problems with this limitation. My only real 'beef' with the limitation is that I'd like to use a 'whole document replacement' simply so I can move the cursor. And that kind of looks like an 'abuse' of a very heavy gun to solve a smaller issue. So I could live with it if the 'I want to move the cursor' issue is addressed more directly.

**Comment by puremourning** [comment]
> Yes, but you would be able to control precisely where the cursor ends up via $0 snippet parameter.

Only if:

- you are using snippets for completion
- the client supports them
- your users like workflows that include snippets

Not all of these are necessarily true. Indeed, often all three are not true.
